### PR TITLE
[SPARK-41177][PROTOBUF][TESTS] Fix maven test failed of `protobuf` module

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Maven and sbt use different protobuf plugin, this causes the contents of `connector/protobuf/src/test/resources/protobuf` to be compiled into different directories in `scala-2.12/test-classes` by different tools:

- Maven : compiled to `connector/protobuf/target/scala-2.12/test-classes/protobuf `
- sbt: compiled to `connector/protobuf/target/scala-2.12/test-classes/`

The files in the `resources/protobuf` directory are read in the test case of the `protobuf` module, which can pass with sbt test,  but will fail in the maven test because the `protobuf/` prefix should be added when using maven.

So this pr add a helper method to passing in two alternative directories to make maven test also successful.

### Why are the changes needed?
Fix maven test failed of `protobuf` module.

### Does this PR introduce _any_ user-facing change?
No, just for test

### How was this patch tested?

- Pass GA
- Manual test

```
mvn clean install -pl connector/protobuf  -am -DskipTests
mvn test -pl connector/protobuf  
```

**Before**

```
ProtobufSerdeSuite:
org.apache.spark.sql.protobuf.ProtobufSerdeSuite *** ABORTED ***
  java.lang.RuntimeException: Unable to load a Suite class org.apache.spark.sql.protobuf.ProtobufSerdeSuite that was discovered in the runpath: null
  at org.scalatest.tools.DiscoverySuite$.getSuiteInstance(DiscoverySuite.scala:80)
  at org.scalatest.tools.DiscoverySuite.$anonfun$nestedSuites$1(DiscoverySuite.scala:38)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
  at scala.collection.Iterator.foreach(Iterator.scala:943)
  at scala.collection.Iterator.foreach$(Iterator.scala:943)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
  at scala.collection.IterableLike.foreach(IterableLike.scala:74)
  at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
  at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
  at scala.collection.TraversableLike.map(TraversableLike.scala:286)
  ...
  Cause: java.lang.NullPointerException:
  at org.apache.spark.sql.test.SQLTestUtilsBase.testFile(SQLTestUtils.scala:466)
  at org.apache.spark.sql.test.SQLTestUtilsBase.testFile$(SQLTestUtils.scala:465)
  at org.apache.spark.sql.protobuf.ProtobufSerdeSuite.testFile(ProtobufSerdeSuite.scala:34)
  at org.apache.spark.sql.protobuf.ProtobufSerdeSuite.<init>(ProtobufSerdeSuite.scala:39)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  at java.lang.Class.newInstance(Class.java:442)
  at org.scalatest.tools.DiscoverySuite$.getSuiteInstance(DiscoverySuite.scala:66)
  ...
ProtobufFunctionsSuite:
org.apache.spark.sql.protobuf.ProtobufFunctionsSuite *** ABORTED ***
  java.lang.RuntimeException: Unable to load a Suite class org.apache.spark.sql.protobuf.ProtobufFunctionsSuite that was discovered in the runpath: null
  at org.scalatest.tools.DiscoverySuite$.getSuiteInstance(DiscoverySuite.scala:80)
  at org.scalatest.tools.DiscoverySuite.$anonfun$nestedSuites$1(DiscoverySuite.scala:38)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
  at scala.collection.Iterator.foreach(Iterator.scala:943)
  at scala.collection.Iterator.foreach$(Iterator.scala:943)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
  at scala.collection.IterableLike.foreach(IterableLike.scala:74)
  at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
  at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
  at scala.collection.TraversableLike.map(TraversableLike.scala:286)
  ...
  Cause: java.lang.NullPointerException:
  at org.apache.spark.sql.test.SQLTestUtilsBase.testFile(SQLTestUtils.scala:466)
  at org.apache.spark.sql.test.SQLTestUtilsBase.testFile$(SQLTestUtils.scala:465)
  at org.apache.spark.sql.protobuf.ProtobufFunctionsSuite.testFile(ProtobufFunctionsSuite.scala:35)
  at org.apache.spark.sql.protobuf.ProtobufFunctionsSuite.<init>(ProtobufFunctionsSuite.scala:39)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  at java.lang.Class.newInstance(Class.java:442)
  at org.scalatest.tools.DiscoverySuite$.getSuiteInstance(DiscoverySuite.scala:66)
  ...
ProtobufCatalystDataConversionSuite:
org.apache.spark.sql.protobuf.ProtobufCatalystDataConversionSuite *** ABORTED ***
  java.lang.RuntimeException: Unable to load a Suite class org.apache.spark.sql.protobuf.ProtobufCatalystDataConversionSuite that was discovered in the runpath: null
  at org.scalatest.tools.DiscoverySuite$.getSuiteInstance(DiscoverySuite.scala:80)
  at org.scalatest.tools.DiscoverySuite.$anonfun$nestedSuites$1(DiscoverySuite.scala:38)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
  at scala.collection.Iterator.foreach(Iterator.scala:943)
  at scala.collection.Iterator.foreach$(Iterator.scala:943)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
  at scala.collection.IterableLike.foreach(IterableLike.scala:74)
  at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
  at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
  at scala.collection.TraversableLike.map(TraversableLike.scala:286)
  ...
  Cause: java.lang.NullPointerException:
  at org.apache.spark.sql.test.SQLTestUtilsBase.testFile(SQLTestUtils.scala:466)
  at org.apache.spark.sql.test.SQLTestUtilsBase.testFile$(SQLTestUtils.scala:465)
  at org.apache.spark.sql.protobuf.ProtobufCatalystDataConversionSuite.testFile(ProtobufCatalystDataConversionSuite.scala:34)
  at org.apache.spark.sql.protobuf.ProtobufCatalystDataConversionSuite.<init>(ProtobufCatalystDataConversionSuite.scala:39)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  at java.lang.Class.newInstance(Class.java:442)
  at org.scalatest.tools.DiscoverySuite$.getSuiteInstance(DiscoverySuite.scala:66)
  ...
Run completed in 717 milliseconds.
Total number of tests run: 0
Suites: completed 1, aborted 3
Tests: succeeded 0, failed 0, canceled 0, ignored 0, pending 0
*** 3 SUITES ABORTED ***
```

**After**

```
[info] ProtobufFunctionsSuite:
[info] - roundtrip in to_protobuf and from_protobuf - struct (4 seconds, 168 milliseconds)
[info] - roundtrip in from_protobuf and to_protobuf - Repeated (570 milliseconds)
[info] - roundtrip in from_protobuf and to_protobuf - Repeated Message Once (340 milliseconds)
[info] - roundtrip in from_protobuf and to_protobuf - Repeated Message Twice (215 milliseconds)
[info] - roundtrip in from_protobuf and to_protobuf - Map (589 milliseconds)
[info] - roundtrip in from_protobuf and to_protobuf - Enum (187 milliseconds)
[info] - roundtrip in from_protobuf and to_protobuf - Multiple Message (346 milliseconds)
[info] - Handle recursive fields in Protobuf schema, A->B->A (117 milliseconds)
[info] - Handle recursive fields in Protobuf schema, C->D->Array(C) (12 milliseconds)
[info] - Handle extra fields : oldProducer -> newConsumer (91 milliseconds)
[info] - Handle extra fields : newProducer -> oldConsumer (7 milliseconds)
[info] - roundtrip in to_protobuf and from_protobuf - with nulls (457 milliseconds)
[info] - from_protobuf filter to_protobuf (356 milliseconds)
[info] - Handle TimestampType between to_protobuf and from_protobuf (560 milliseconds)
[info] - Handle DayTimeIntervalType between to_protobuf and from_protobuf (547 milliseconds)
[info] - raise cannot construct protobuf descriptor error (13 milliseconds)
[info] ProtobufCatalystDataConversionSuite:
[info] - single StructType(StructField(int32_type,IntegerType,true)) with seed 989 (76 milliseconds)
[info] - single StructType(StructField(double_type,DoubleType,true)) with seed 65 (51 milliseconds)
[info] - single StructType(StructField(float_type,FloatType,true)) with seed 43 (32 milliseconds)
[info] - single StructType(StructField(bytes_type,BinaryType,true)) with seed 657 (62 milliseconds)
[info] - single StructType(StructField(string_type,StringType,true)) with seed 965 (43 milliseconds)
[info] - Handle unsupported input of message type (31 milliseconds)
[info] - filter push-down to Protobuf deserializer (23 milliseconds)
[info] - ProtobufDeserializer with binary type (1 millisecond)
[info] - Full names for message using descriptor file (1 millisecond)
[info] ProtobufSerdeSuite:
[info] - Test basic conversion (7 milliseconds)
[info] - Fail to convert with field type mismatch (3 milliseconds)
[info] - Fail to convert with missing nested Protobuf fields for serializer (2 milliseconds)
[info] - Fail to convert with deeply nested field type mismatch (8 milliseconds)
[info] - Fail to convert with missing Catalyst fields (3 milliseconds)
[info] - raise cannot parse and construct protobuf descriptor error (2 milliseconds)
[info] Run completed in 12 seconds, 167 milliseconds.
[info] Total number of tests run: 31
[info] Suites: completed 3, aborted 0
[info] Tests: succeeded 31, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### CODE REVIEW

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala [hunk 1]
```diff
@@ -34,9 +34,10 @@ import org.apache.spark.unsafe.types.UTF8String
 class ProtobufCatalystDataConversionSuite
     extends SparkFunSuite
     with SharedSparkSession
-    with ExpressionEvalHelper {
+    with ExpressionEvalHelper
+    with ProtobufTestBase {
 
-  private val testFileDesc = testFile("catalyst_types.desc").replace("file:/", "/")
+  private val testFileDesc = testFile("catalyst_types.desc", "protobuf/catalyst_types.desc")
```
Comment (LuciferYang) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala:40]: sbt need `catalyst_types.desc` but maven need `protobuf/catalyst_types.desc`

**Comment by LuciferYang** [comment]
cc @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by LuciferYang** [comment]
Thanks @HyukjinKwon

**Pull request closed without merging**